### PR TITLE
Fixed point operations

### PR DESCRIPTION
**Pull request by VoR0220**
This contains all operations EXCEPT for multiplication and exponentiation (which have been marked as unusable for the time being). I think it would be best to view this as the last to merge if anything (fix up #1079 and make sure #1080 passes) then merge this in. In the meantime, work will be continued on the multiplication assembly.

### CODE REVIEW

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -1350,11 +1350,28 @@ void ExpressionCompiler::appendOrdinaryBinaryOperatorCode(Token::Value _operator
...
+	{
+		FixedPointType const& type = dynamic_cast<FixedPointType const&>(_type);
+		isSigned = type.isSigned();		
+		isFractional = true;
+		fractionalBits = type.fractionalBits();
+		fractionShift = (u256(1) << (fractionalBits));
+		halfShift = (u256(1) << (fractionalBits / 2));
+	}
+	else
+		BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment("Expected numeric type."));
```
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Can use `solAssert(false,..`)

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 2]
```diff
@@ -1365,6 +1382,8 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
 		m_context << Instruction::SUB;
 		break;
 	case Token::Mul:
+		if (isFractional)
+			solAssert(false, "Multiplication not yet implemented - FixedPointType.");
```
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Should be `solUnimplemented`

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 3]
```diff
@@ -1375,9 +1394,23 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
...
+				m_context << (isSigned ? Instruction::SDIV : Instruction::DIV) << fractionShift << Instruction::MUL;
+				//now redo the process...
+				m_context << Instruction::DUP2 << Instruction::DUP4;
+				//but this time, get the fraction portion.
+				m_context << fractionShift << Instruction::SWAP1 << Instruction::MUL << (isSigned ? Instruction::SDIV : Instruction::DIV);
+				//add
+				m_context << Instruction::ADD;
+			}
+ 			else
+				m_context << (isSigned ? Instruction::SDIV : Instruction::DIV);
```
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp]: This is shared with the `isFractional` case. It could be moved to the top of the block.

Comment (VoR0220) [reply libsolidity/codegen/ExpressionCompiler.cpp]: True it could.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 4]
```diff
@@ -8553,6 +8553,119 @@ BOOST_AUTO_TEST_CASE(inline_array_rationals_return)
...
+				return a;
+			}
+			function decPost() returns (ufixed a) {
+				a = 4.5;
+				a--;
+				return a;
+			}
+			function daWhile() returns (ufixed8x8) {
+				var i = 3.25;
+				while ( i < 5.5) {
```
Comment (axic) [inline test/libsolidity/SolidityEndToEndTest.cpp:8580]: Extra space there.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 5]
```diff
@@ -1334,11 +1350,28 @@ void ExpressionCompiler::appendOrdinaryBinaryOperatorCode(Token::Value _operator
 
 void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Type const& _type)
 {
-	IntegerType const& type = dynamic_cast<IntegerType const&>(_type);
-	bool const c_isSigned = type.isSigned();
+	bool isSigned = false;
+	bool isFractional = false;
+	u256 fractionShift = u256(1);
```
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp:1355]: It seems that in this PR only `isSigned` and `isFractional` is used.

**Comment by pirapira** [comment]
For the same reason as #1080, this pull-request seems to need changing.

**Pull request closed without merging**